File: solutions/scala/154-all-nodes-distance-k-in-binary-tree.scala
/**
 * Problem 154: All Nodes Distance K in Binary Tree (LeetCode 863)
 * Difficulty: Med
 * Language: Scala
 */
def distanceK(adj: Map[Int, List[Int]], target: Int, k: Int): List[Int] = {
  def bfs(queue: List[(Int, Int)], visited: Set[Int]): List[Int] = queue match {
    case Nil => Nil
    case (node, dist) :: rest if dist == k =>
      node :: bfs(rest, visited)
    case (node, dist) :: rest =>
      val neighbors = adj.getOrElse(node, Nil).filterNot(visited.contains)
      bfs(rest ++ neighbors.map((_, dist + 1)), visited ++ neighbors)
  }
  bfs(List((target, 0)), Set(target))
}

File: solutions/rust/154-all-nodes-distance-k-in-binary-tree.rs
/**
 * Problem 154: All Nodes Distance K in Binary Tree (LeetCode 863)
 * Difficulty: Med
 * Language: Rust
 */
use std::collections::{HashMap, HashSet, VecDeque};
fn distance_k(adj: &HashMap<i32, Vec<i32>>, target: i32, k: i32) -> Vec<i32> {
    let mut visited = HashSet::new();
    let mut queue = VecDeque::new();
    queue.push_back((target, 0));
    visited.insert(target);
    let mut result = vec![];
    while let Some((node, dist)) = queue.pop_front() {
        if dist == k { result.push(node); }
        if dist < k {
            if let Some(neighbors) = adj.get(&node) {
                for &n in neighbors.iter().filter(|n| !visited.contains(n)) {
                    visited.insert(n);
                    queue.push_back((n, dist + 1));
                }
            }
        }
    }
    result
}
